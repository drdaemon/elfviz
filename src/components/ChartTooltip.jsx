import React, { Component } from 'react';
import PropTypes from 'prop-types';

class ChartTooltip extends Component {
  render() {
    let visibility = 'hidden';
    let transform = '';
    let width = 200;
    let height = 54;
    let transformText = 'translate(' + width / 2 + ',' + (height / 2 - 10) + ')';
    let transformArrow = '';

    if (this.props.display) {
      const { x, y } = this.props.pos;
      visibility = 'visible';

      if (y > height) {
        transform = 'translate(' + (x - width / 2) + ',' + (y - height - 20) + ')';
        transformArrow = 'translate(' + (width / 2 - 20) + ',' + (height - .2) + ')';
      } else if (y < height) {
        transform = 'translate(' + (x - width / 2) + ',' + (Math.round(y) + 20) + ')';
        transformArrow = 'translate(' + (width / 2 - 20) + ',' + 0 + ') rotate(180,20,0)';
      }
    }

    const text1 = this.props.xValue + ': ' + this.props.data.key;
    const text2 = this.props.yValue + ': ' + this.props.data.value;
    return (
      <g transform={transform}>
        <rect class={this.props.bgStyle} is width={width} height={height} rx="5" ry="5" visibility={visibility} />
        <polygon class={this.props.bgStyle} is points="10,0  30,0  20,10" transform={transformArrow} visibility={visibility} />
        <text is visibility={visibility} transform={transformText}>
          <tspan is x="0" class={this.props.textStyle1} text-anchor="middle">{text1}</tspan>
          <tspan is x="0" class={this.props.textStyle2} text-anchor="middle" dy="25">{text2}</tspan>
        </text>
      </g>
    );
  }
}

ChartTooltip.propTypes = {
  data: PropTypes.object.isRequired,
  pos: PropTypes.object,
  tooltip: PropTypes.object,
  bgStyle: PropTypes.string,
  textStyle1: PropTypes.string,
  textStyle2: PropTypes.string,
  xValue: PropTypes.string,
  yValue: PropTypes.string,
};

export default ChartTooltip;
