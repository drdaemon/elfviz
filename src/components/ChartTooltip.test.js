import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import ChartTooltip from './ChartTooltip';

const data = {
  key: 'key', value: 'value',
}

describe('ChartTooltip', () => {
  it('renders without crashing', () => {
    shallow(<ChartTooltip data={data} />);
  });

  it('should match an empty snapshot', () => {
    const wrapper = shallow(<ChartTooltip data={data} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should match snapshot with provided data', () => {
    const wrapper = shallow(<ChartTooltip
      data={data}
      pos={{ x: 100, y: 20 }}
      textStyle1="style1"
      textStyle2="style2"
      bgStyle="tooltip-bg"
      xValue="xValue"
      yValue="yValue"
    />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should match snapshot with provided data and visible', () => {
    const wrapper = shallow(<ChartTooltip
      display
      data={data}
      pos={{ x: 100, y: 20 }}
      textStyle1="style1"
      textStyle2="style2"
      bgStyle="tooltip-bg"
      xValue="xValue"
      yValue="yValue"
    />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should match snapshot with provided data and visible when y > height', () => {
    const wrapper = shallow(<ChartTooltip
      display
      data={data}
      pos={{ x: 100, y: 200 }}
      textStyle1="style1"
      textStyle2="style2"
      bgStyle="tooltip-bg"
      xValue="xValue"
      yValue="yValue"
    />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

});
