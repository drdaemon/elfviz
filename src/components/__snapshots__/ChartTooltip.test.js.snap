// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartTooltip should match an empty snapshot 1`] = `
<g
  transform=""
>
  <rect
    height={54}
    is={true}
    rx="5"
    ry="5"
    visibility="hidden"
    width={200}
  />
  <polygon
    is={true}
    points="10,0  30,0  20,10"
    transform=""
    visibility="hidden"
  />
  <text
    is={true}
    transform="translate(100,17)"
    visibility="hidden"
  >
    <tspan
      is={true}
      text-anchor="middle"
      x="0"
    >
      undefined: key
    </tspan>
    <tspan
      dy="25"
      is={true}
      text-anchor="middle"
      x="0"
    >
      undefined: value
    </tspan>
  </text>
</g>
`;

exports[`ChartTooltip should match snapshot with provided data 1`] = `
<g
  transform=""
>
  <rect
    class="tooltip-bg"
    height={54}
    is={true}
    rx="5"
    ry="5"
    visibility="hidden"
    width={200}
  />
  <polygon
    class="tooltip-bg"
    is={true}
    points="10,0  30,0  20,10"
    transform=""
    visibility="hidden"
  />
  <text
    is={true}
    transform="translate(100,17)"
    visibility="hidden"
  >
    <tspan
      class="style1"
      is={true}
      text-anchor="middle"
      x="0"
    >
      xValue: key
    </tspan>
    <tspan
      class="style2"
      dy="25"
      is={true}
      text-anchor="middle"
      x="0"
    >
      yValue: value
    </tspan>
  </text>
</g>
`;

exports[`ChartTooltip should match snapshot with provided data and visible 1`] = `
<g
  transform="translate(0,40)"
>
  <rect
    class="tooltip-bg"
    height={54}
    is={true}
    rx="5"
    ry="5"
    visibility="visible"
    width={200}
  />
  <polygon
    class="tooltip-bg"
    is={true}
    points="10,0  30,0  20,10"
    transform="translate(80,0) rotate(180,20,0)"
    visibility="visible"
  />
  <text
    is={true}
    transform="translate(100,17)"
    visibility="visible"
  >
    <tspan
      class="style1"
      is={true}
      text-anchor="middle"
      x="0"
    >
      xValue: key
    </tspan>
    <tspan
      class="style2"
      dy="25"
      is={true}
      text-anchor="middle"
      x="0"
    >
      yValue: value
    </tspan>
  </text>
</g>
`;

exports[`ChartTooltip should match snapshot with provided data and visible when y > height 1`] = `
<g
  transform="translate(0,126)"
>
  <rect
    class="tooltip-bg"
    height={54}
    is={true}
    rx="5"
    ry="5"
    visibility="visible"
    width={200}
  />
  <polygon
    class="tooltip-bg"
    is={true}
    points="10,0  30,0  20,10"
    transform="translate(80,53.8)"
    visibility="visible"
  />
  <text
    is={true}
    transform="translate(100,17)"
    visibility="visible"
  >
    <tspan
      class="style1"
      is={true}
      text-anchor="middle"
      x="0"
    >
      xValue: key
    </tspan>
    <tspan
      class="style2"
      dy="25"
      is={true}
      text-anchor="middle"
      x="0"
    >
      yValue: value
    </tspan>
  </text>
</g>
`;
